tests.ts

```diff
@@ -0,0 +1,2093 @@
+import { assertEquals, assertNotEquals, assert } from "jsr:@std/assert";
+import * as path from "jsr:@std/path";
+// import * as fs from "jsr:@std/fs";
+import { exists } from "jsr:@std/fs";
+import * as yaml from "jsr:@std/yaml";
+import * as toml from "jsr:@std/toml";
+
+import process from 'node:process';
+import http from 'node:http';
+import https from 'node:https';
+import { URL } from 'node:url';
+import { Buffer } from "node:buffer";
+import os from 'node:os';
+import fs from 'node:fs';
+
+
+// UTILITIES
+
+const ENV_VAR_REGISTRY: string = 'WASMER_REGISTRY';
+const ENV_VAR_NAMESPACE: string = 'WASMER_NAMESPACE';
+const ENV_VAR_TOKEN: string = 'WASMER_TOKEN';
+const ENV_VAR_APP_DOMAIN: string = 'WASMER_APP_DOMAIN';
+const ENV_VAR_EDGE_SERVER: string = 'EDGE_SERVER';
+const ENV_VAR_WASMER_PATH: string = 'WASMER_PATH';
+const ENV_VAR_WASMOPTICON_DIR: string = 'WASMOPTICON_DIR';
+
+const REGISTRY_DEV: string = 'https://registry.wasmer.wtf/graphql';
+const REGISTRY_PROD: string = 'https://registry.wasmer.io/graphql';
+
+const appDomainMap = {
+  [REGISTRY_PROD]: 'wasmer.app',
+  [REGISTRY_DEV]: 'wasmer.dev',
+};
+
+const DEFAULT_NAMESPACE: string = 'wasmer-integration-tests';
+
+async function sleep(milliseconds: number) {
+  return new Promise((resolve) => setTimeout(resolve, milliseconds));
+}
+
+function randomAppName(): string {
+  const id = crypto.randomUUID();
+  return 't-' + id.replace(/\-/g, '').substr(0, 20);
+}
+
+// Path to the wasmopticon repo.
+async function wasmopticonDir(): Promise<string> {
+  const WASMOPTICON_GIT_URL = 'https://github.com/wasix-org/wasmopticon.git';
+  let dir = process.env[ENV_VAR_WASMOPTICON_DIR];
+  if (dir) {
+    const doesExist = await exists(dir);
+    if (!doesExist) {
+      throw new Error(`${ENV_VAR_WASMOPTICON_DIR} is set, but directory does not exist: ${dir}`);
+    }
+    return dir;
+  }
+
+  // No env var set, check the default location.
+  const localDir = path.join(process.cwd(), 'wasmopticon');
+  const doesExist = await exists(localDir);
+  if (doesExist) {
+    return localDir;
+  }
+
+
+  console.log('wasmopticon dir not found')
+  console.log(`Cloning ${WASMOPTICON_GIT_URL} to ${localDir}...`);
+  const cmd = new Deno.Command('git', {
```

Comment:
cloning submodule causes race condition in tests, so the CI's should be configured in a way that testing starts after submodule is cloned

Reply:
we should keep that in mind when updating the ci and keep the submodule cloning